server/src/internalClusterTest/java/org/elasticsearch/repositories/IndexSnapshotsServiceIT.java

```diff
@@ -0,0 +1,358 @@
+
+            final Map<String, RepositoryException> failures = response.getRepositoryFailures();
+            for (String repository : repositories) {
+                RepositoryException repositoryException = failures.get(repository);
+                assertThat(repositoryException, is(notNullValue()));
+                assertThat(
+                    repositoryException.getMessage(),
+                    equalTo(String.format(Locale.ROOT, "[%s] Unable to find the latest snapshot for shard [[idx][0]]", repository))
+                );
+            }
+        } else {
+            expectThrows(RepositoryException.class, responseFuture::actionGet);
+        }
+
+        disableRepoConsistencyCheck("This test checks an empty repository");
+    }
+
+    public void testGetShardSnapshotFromEmptyRepositoryReturnsEmptyResult() {
+        final String fsRepoName = randomAlphaOfLength(10);
+        createRepository(fsRepoName, FsRepository.TYPE);
+
+        final Optional<ShardSnapshotInfo> indexShardSnapshotInfo = getLatestSnapshotForShard(fsRepoName, "test", 0);
+        assertThat(indexShardSnapshotInfo.isEmpty(), equalTo(true));
+
+        disableRepoConsistencyCheck("This test checks an empty repository");
+    }
+
+    public void testGetShardSnapshotFromUnknownIndexReturnsEmptyResult() {
+        final String fsRepoName = randomAlphaOfLength(10);
+        createRepository(fsRepoName, FsRepository.TYPE);
+
+        createSnapshot(fsRepoName, "snap-1", Collections.emptyList());
+
+        final Optional<ShardSnapshotInfo> indexShardSnapshotInfo = getLatestSnapshotForShard(fsRepoName, "test", 0);
+        assertThat(indexShardSnapshotInfo.isEmpty(), equalTo(true));
+    }
+
+    public void testGetShardSnapshotFromUnknownShardReturnsEmptyResult() {
+        final String fsRepoName = randomAlphaOfLength(10);
+        final String indexName = "test-idx";
+
+        createIndexWithContent(indexName);
+
+        createRepository(fsRepoName, FsRepository.TYPE);
+        createSnapshot(fsRepoName, "snap-1", Collections.singletonList(indexName));
+
+        final Optional<ShardSnapshotInfo> indexShardSnapshotInfo = getLatestSnapshotForShard(fsRepoName, indexName, 100);
+        assertThat(indexShardSnapshotInfo.isEmpty(), equalTo(true));
+    }
+
+    public void testGetShardSnapshotOnEmptyRepositoriesListThrowsAnError() {
+        final ActionFuture<GetShardSnapshotResponse> response = getLatestSnapshotForShardFuture(Collections.emptyList(), "idx", 0, false);
+
+        expectThrows(ActionRequestValidationException.class, response::actionGet);
+    }
+
+    public void testGetShardSnapshotReturnsTheLatestSuccessfulSnapshot() throws Exception {
+        final String repoName = "repo-name";
+        final Path repoPath = randomRepoPath();
+        createRepository(repoName, FsRepository.TYPE, repoPath);
+
+        final boolean useBwCFormat = randomBoolean();
+        if (useBwCFormat) {
+            final Version version = randomFrom(
```

Comment:
You could probably just use `org.elasticsearch.test.VersionUtils#randomVersionBetween` here to get a little more coverage of the in-between versions and just use any random version between current and 7.5.0 (old snapshot format) randomly?